src/utils/crypto_utils/src/crypto_lib.c

```diff
@@ -441,3 +458,139 @@ void CryptoUtils_FreeCryptoKeyHandle(CryptoKeyHandle key)
+ * @return CONSTBUFFER_HANDLE The digest buffer handle. Caller will own it and need to call CONSTBUFFER_DecRef() to free.
+ */
+CONSTBUFFER_HANDLE CryptoUtils_CreateSha256Hash(const CONSTBUFFER_HANDLE buf)
+{
+    const size_t Sha256DigestNumBytes = 32;
+    unsigned char hashBuf[Sha256DigestNumBytes];
+    memset(hashBuf, 0, sizeof(hashBuf));
+    const CONSTBUFFER* constbuf = CONSTBUFFER_GetContent(buf);
+
+    CONSTBUFFER_HANDLE out_handle = NULL;
+
+    unsigned char* output_digest = NULL;
+    const EVP_MD* msg_digest = NULL;
+    EVP_MD_CTX* mdctx;
+
+    if (buf == NULL)
+    {
+        return NULL;
+    }
+
+    // Initialize OpenSSL digest context
+    if ((mdctx = EVP_MD_CTX_new()) == NULL)
+    {
+        goto done;
+    }
+
+    msg_digest = GetEvpMdFromShaAlg("sha256");
+
+    // Initialize OpenSSL digest operation
+    if (1 != EVP_DigestInit_ex(mdctx, msg_digest, NULL))
+    {
+        goto done;
+    }
+
+    // Hash the input data
+    if (1 != EVP_DigestUpdate(mdctx, constbuf->buffer, constbuf->size))
+    {
+        goto done;
+    }
+
+    // Finalize the digest and get the output
+    if (1 != EVP_DigestFinal_ex(mdctx, &hashBuf[0], NULL))
+    {
+        goto done;
+    }
+
+    output_digest = (unsigned char*)malloc(Sha256DigestNumBytes);
+    if (output_digest != NULL)
+    {
+        memcpy(output_digest, &hashBuf[0], Sha256DigestNumBytes);
+
+        out_handle = CONSTBUFFER_Create(output_digest, Sha256DigestNumBytes);
+        output_digest = NULL;
+    }
+
+done:
+    EVP_MD_CTX_free(mdctx);
+    free(output_digest);
+
+    return out_handle;
+}
+
+CONSTBUFFER_HANDLE CryptoUtils_GeneratePublicKey(const char* modulus_b64url, const char* exponent_b64url)
+{
+    CONSTBUFFER_HANDLE publicKeyData = NULL;
+
+    unsigned char *modulus_bytes, *exponent_bytes;
+    int modulus_length, exponent_length;
+    BIGNUM* bn_modulus = NULL;
+    BIGNUM* bn_exponent = NULL;
+    RSA* rsa = NULL;
+    unsigned char* der_encoded_bytes = NULL;
+    int der_length = 0;
+
+    modulus_length = Base64URLDecode(modulus_b64url, &modulus_bytes);
+    if (modulus_length == 0)
+    {
+        goto done;
+    }
+
+    exponent_length = Base64URLDecode(exponent_b64url, &exponent_bytes);
+    if (exponent_length == 0)
+    {
+        goto done;
+    }
+
+    bn_modulus = BN_bin2bn(modulus_bytes, modulus_length, NULL);
+    if (bn_modulus == NULL)
+    {
+        goto done;
+    }
+
+    bn_exponent = BN_bin2bn(exponent_bytes, exponent_length, NULL);
+    if (bn_exponent == NULL)
+    {
+        goto done;
+    }
+
+    rsa = RSA_new();
+    if (rsa == NULL)
+    {
+        goto done;
+    }
+
+    if (RSA_set0_key(rsa, bn_modulus, bn_exponent, NULL) == 0)
+    {
+        goto done;
+    }
+
+    der_encoded_bytes = NULL;
+    der_length = i2d_RSAPublicKey(rsa, &der_encoded_bytes); // DER PKCS#1
+    if (der_length == 0)
+    {
+        goto done;
+    }
+    // der_length = i2d_RSA_PUBKEY(rsa, &der_encoded_bytes); // cert format?
```

Comment:
Is this a TODO?

Reply:
will remove in follow-up pr

Reply:
Fixed in upstream topic branch: https://github.com/Azure/iot-hub-device-update/pull/489/commits/ac88823edb927123e4462b31ddd21dbf651e4e91